File: ecos-records/src/main/java/ru/citeck/ecos/records2/meta/RecordsTemplateService.java
package ru.citeck.ecos.records2.meta;

import ru.citeck.ecos.commons.utils.TmplUtils;
import ru.citeck.ecos.records3.record.atts.dto.RecordAtts;
import ru.citeck.ecos.records2.RecordRef;
import ru.citeck.ecos.records3.RecordsService;
import ru.citeck.ecos.records3.RecordsServiceFactory;

import java.util.Set;

public class RecordsTemplateService {

    private final RecordsService recordsService;

    public RecordsTemplateService(RecordsServiceFactory recordsServiceFactory) {
        recordsService = recordsServiceFactory.getRecordsServiceV1();
    }

    public <T> T resolve(T template, RecordRef recordRef) {

        if (template == null || recordRef == null) {
            return template;
        }

        Set<String> atts = TmplUtils.getAtts(template);
        if (atts.isEmpty()) {
            return template;
        }

        RecordAtts meta = recordsService.getAtts(recordRef, atts);

        return TmplUtils.applyAtts(template, meta.getAtts());
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records2/meta/RecordsTemplateService.kt
package ru.citeck.ecos.records2.meta

import ru.citeck.ecos.commons.utils.TmplUtils
import ru.citeck.ecos.records2.ServiceFactoryAware
import ru.citeck.ecos.records3.RecordsService
import ru.citeck.ecos.records3.RecordsServiceFactory
import ru.citeck.ecos.records3.record.atts.value.AttValueCtx
import ru.citeck.ecos.records3.record.atts.value.RecordAttValueCtx
import ru.citeck.ecos.webapp.api.entity.EntityRef

class RecordsTemplateService : ServiceFactoryAware {

    private lateinit var recordsService: RecordsService

    fun <T> resolve(template: T, recordRef: EntityRef?): T {
        return if (EntityRef.isEmpty(recordRef)) {
            template
        } else {
            resolve(template, recordRef as Any?)
        }
    }

    fun <T> resolve(template: T, record: Any?): T {
        if (template == null || record == null) {
            return template
        }
        val atts = TmplUtils.getAtts(template)
        if (atts.isEmpty()) {
            return template
        }
        val valueCtx = if (record is AttValueCtx) {
            record
        } else {
            RecordAttValueCtx(record, recordsService)
        }
        val attsRes = valueCtx.getAtts(atts)
        return TmplUtils.applyAtts(template, attsRes)
            ?: error("Apply atts can't be performed. Template: $template Record: $record")
    }

    override fun setRecordsServiceFactory(serviceFactory: RecordsServiceFactory) {
        this.recordsService = serviceFactory.recordsService
    }
}
